#### Loading the MNIST dataset with Keras
Classify grayscale images of handwritten digits (28x28 pixels) into 10 categories (0 - 9). <br>
*(Building and training a neutral network to classify handwritten digits in less than 20 lines of Python code)*

In [1]:
from keras.datasets import mnist

In [13]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [3]:
# look at the training data
train_images.shape

(60000, 28, 28)

*training_images* is stored in an array of shape (600000, 28, 28) of type uint8 with values in the [0,255] interval.

In [4]:
len(train_labels)

60000

In [5]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

#### The network architecture

In [6]:
from keras import models
from keras import layers

In [7]:
# network consists of sequence of 2 Dense layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

Second layer is a 10-way softmax layer, which will return an array of 10 probabilty scores (summing to 1). <br>
Each score will be the probability that the current image belongs to one of the 10 digit classes.

#### The compilation step

In [19]:
network.compile(optimizer = 'rmsprop',
               loss = 'categorical_crossentropy',
               metrics = ['accuracy'])

In [15]:
# transform train_images into float32 array of shape (60000, 28 * 28) with values between 0 and 1.
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

#### Preparing the labels

In [17]:
from tensorflow.keras.utils import to_categorical
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [20]:
# fit model to the training data
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 2s 3ms/step - loss: 1.4817 - accuracy: 0.6574
Epoch 2/5
469/469 [==============================] - 2s 3ms/step - loss: 0.5904 - accuracy: 0.8533
Epoch 3/5
469/469 [==============================] - 2s 3ms/step - loss: 0.4131 - accuracy: 0.8887
Epoch 4/5
469/469 [==============================] - 2s 3ms/step - loss: 0.3543 - accuracy: 0.9004
Epoch 5/5
469/469 [==============================] - 2s 4ms/step - loss: 0.3254 - accuracy: 0.9070


*loss* of the network and *accuracy* is displayed <br>
We reach 90.7% accuracy.

In [21]:
# Check that the model performs well on test set too
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

313/313 [==============================] - 0s 983us/step - loss: 49.5265 - accuracy: 0.8255
test_acc: 0.8255000114440918
